src/main/scala/strawman/collection/immutable/Set.scala

```diff
@@ -22,4 +23,15 @@ trait SetMonoTransforms[A, +Repr]
 
   def - (elem: A): Repr
 
+}
+
+trait SetPolyTransforms[A, +C[X] <: Set[X] with SetLike[X, C]]
+  extends collection.SetPolyTransforms[A, C] {
+
+  protected def coll: C[A]
+
+  def ++ (that: collection.Set[A]): C[A] =
+    if (that.isEmpty) coll
+    else that.foldLeft[C[A]](coll)(_ + _)
```

Comment:
For reference: I tried inlining the `foldLeft` loop and only got a very slight improvement. On the other hand, removing the `if (that.isEmpty)` check had more impact.

BTW, in the current collections, the implementation is just a `foldLeft` (not inlined, and without first checking if the collection to concatenate is empty or not).